### DISCUSSION TITLE
Simpler, more consistent error handling

### DISCUSSION THREAD

**Discussion by Rich-Harris** [sveltejs/kit#6499] [Announcements]
_Edited following discussion in https://github.com/sveltejs/kit/discussions/6499#discussioncomment-3527390_

SvelteKit handles errors very inconsistently. If you throw an `Error` in a `load` function, `$page.error` will be that very same `Error` object during server-side rendering, but a POJO (created using [somewhat opinionated/convoluted logic](https://github.com/sveltejs/kit/blob/eb24cf5f6fa8331316b61bab1c7daf15f207b182/packages/kit/src/runtime/server/utils.js#L35-L73)) in the client:

```svelte
<script>
  import { page } from '$app/stores';

  console.log('$page.error', $page.error instanceof Error); // who knows?!
</script>
```

Another inconsistency: stack traces are included during development, but not in production (because stack traces aren't the sorts of things you should generally make public). But unless you're aware of that distinction, this can just lead to unexpected behaviour.

Meanwhile, error _messages_ (in the case where the error was unexpected, perhaps emanating from the bowels of some library you're using) are also best kept hidden from public view, yet we send them to the client.

All of this confusion extends to JSON representations of these errors. An unexpected error will be an object with a `name`, `message` and (in dev) `stack` plus whatever additional properties happened to be lying around on the original error (e.g. `code`, `frame` etc) while _expected_ errors (the kind thrown with `throw error(status, message)` have `status` (redundantly), `message` and `__is_http_error` (shudder).

Everything would get so much simpler (both implementation-wise and for users) if we eliminated all this ambiguity:

## `handleError` returns a POJO

Unexpected errors already go through `handleError`, which by default prints the error (complete with sourcemapped stacktrace) to the terminal. That seems like plenty — we don't _also_ need to print the stack trace to the `+error.svelte` page or include it in a JSON response, especially when doing so creates all the confusion described above. 

`handleError` could be given the responsibility of creating a JSON-friendly non-sensitive-data-containing object representing the error. It would default to this, if no `handleFile` was specified or no `return` value was provided...

```js
export function handleError({ event, error }) {
  console.error(error.stack);
  return {
    message: 'Internal Error'
  };
}
```

...but could include additional metadata:

```js
export function handleError({ event, error }) {
  const { id, traceid } = sendToLoggingService(error);

  return {
    message: 'Something went wrong!',
    metadata: { id, traceid }
  };
}
```

## Make `$page.error` that same POJO

Instead of playing 'is it an `Error` or `HttpError` or a POJO?', we can make `$page.error` just be a POJO:

```html
<!-- src/routes/+error.svelte -->
<h1>{$page.error.message}</h1>
<p>Our team has been notified. Internal error ID: {$page.error.metadata?.id ?? 'unknown'}</p>
```

For unexpected errors, we default to replacing `message` with 'Internal Error' (not 'Internal Server Error', since the same code could throw an unexpected error when it runs on the server _or_ in the browser). If no message is provided (e.g. `throw error(404)`), we default to just 'Error'.

## Throw POJOs with `error(...)`

When you throw an _expected_ error with `throw error(...)`, the error doesn't go via `handleError`, because that exists primarily to help you track down _unexpected_ errors. But you still might want to have a richer error object than just a `message`, for example containing a `code` that can be used for i18n:

```js
throw error(404, {
  "message": "Not found",
  code: 'NOT_FOUND'
});
```

## Simpler JSON representation

If you hit a route with `fetch`, the response should just be what was returned from `handleError` or thrown with `throw error(...)`:

```js
500
{
  "message": "Something went wrong",
  "metadata": {
    "id": "abc123",
    "traceid": "def456"
  }
}
```

```js
518
{
  "message": "I'm a teapot",
  "code": "I_AM_A_TEAPOT"
}
```

## Type safety

We could type `handleError` and `$page.error` by defining an `App.PageError` interface, as we do with [`App.Locals`](https://kit.svelte.dev/docs/types#app) and so on:

```ts
declare namespace App {
  export interface PageError {
    message: string;
    metadata: {
      id: string;
      traceid: string;
    }
  }
}
```

One open question is whether `message: string` would always be a required property. I think it probably needs to be, otherwise the default behaviour of `handleError` would violate the type constraint. In other words, SvelteKit's types would include this declaration...

```ts
declare namespace App {
  export interface PageError {
    message: string;
  }
}
```

...and app-declared types wouldn't be able to unset it (because that's not how interface merging works).

## In future: better client-side error handling

Unexpected errors that happened during client-side navigation would be printed to the console with `console.error`. Sentry and similar services can deal with this sort of thing, but it might be even nicer if we had some client-side error handling.

In the case of expected client-side errors like the `NOT_FOUND` one above, `$page.error` would just be that object:

```js
console.log($page.error); // { message: 'Not found', code: 'NOT_FOUND' }
```

In the case of _un_expected errors, `$page.error` would just be a `{ message: 'Error' }` object, unless we can find a good way to design a client-side version of `handleError`. My first thought was that it could be a lifecycle function, but there really needs to be a single one that's present for the life of the app, rather than something many components could add. (Update: some thoughts on this below: https://github.com/sveltejs/kit/discussions/6499#discussioncomment-3540775)

**Comment by benmccann** [comment]
I agree that having `error` be an `Error` or `HttpError` or a `POJO` is confusing. I might choose to make it be a `POJO` rather than a `string` though. At the very least it would then be extensible in the future even if there were nothing else we decided to put there now

I used to use Play Framework a lot and one nice feature that had was a randomly generated error ID. You could include it in your error page, which was nice because support requests would often have details like "I keep getting error pages. I just got it with error ae02ef", which we could search the logs for and find rather easily.

`status` is nice on the initial page load, but doesn't really make sense for client-side navigations. I wouldn't mind having it populated only for server-side rendered pages, but won't complain too much if we leave it out. I also think including the `stack` is nice for server-rendered pages and will advocate for that, but know others feel differently about it and so I don't necessarily expect it to make the cut.

**Reply by Rich-Harris** [reply]
Yeah, `$page.error.message` works too. It's actually more consistent with the JSON representation too.

> `status` is nice on the initial page load, but doesn't really make sense for client-side navigations. I wouldn't mind having it populated only for server-side rendered pages, but won't complain too much if we leave it out

It's definitely an abstraction leak, but stuff like this is too useful to disallow it on grounds of purity:

```svelte
{#if $page.status === 401}
  <p>You must be <a href="/login?redirectTo={$page.url.pathname}">logged in</a> to view this page</p>
{:else if $page.status === 404}
  <p>This page doesn't exist! Try the <a href="/">homepage</a></p>
{:else}
  <p>Oh no! Something went wrong. Our team has been notified</p>
{/if}
```

**Reply by Rich-Harris** [reply]
(Also, `$page.error.message` means this change will be non-breaking for most people)

**Reply by Rich-Harris** [reply]
For consistency, `%sveltekit.error%` should be changed to `%sveltekit.error.message%` I think

**Comment by vekunz** [comment]
Returning just one string as an error is not really usable for us. We have fairly complex error handling and need more information than just an error message. More precisely, an error message is the only thing, we do NOT need. Because we do client-side localization, we return an error ID. For deeper debugging, we also need to return a [Trace ID](https://www.w3.org/TR/trace-context/). We show this Trace ID in the UI in several cases (especially unexcepted errors), so that an end user can send us this trace id, and we can find the exact log lines of this call.
We already have to return a JSON string as an error message to return multiple information to our +error.svelte. And in +page.server.ts files we use the feature to return multiple errors to return multiple information of one error.

**Reply by Rich-Harris** [reply]
You'd still have access to the actual error object in `handleError`, and you could capture it via `console.error` or the hypothetical `onLoadError` function. Aside from showing the ID in the UI, does that meet your needs?

I'm not sure what a sensible way to get `$page.error.id` (or similar) to be available is — any thoughts? Ideally we don't just want to convert the entire error object to a POJO like we currently do, since it's pretty weird and unexpected behaviour, and it leaks potentially private information by default. I'm also curious about how you're dealing with these error IDs in the client?

**Reply by vekunz** [reply]
I think it is best to describe our error flow a bit more precisely.

---

**Our current state**

Every endpoint (any load-function, POST, PUT, ...) has a big try-catch in it, to catch any error that is thrown in our server-side code. In the catch-statement, we call an error mapper function. This error mapper maps the catched error (this can be an explicitly thrown custom error or any unexpected error) to three pieces of information, we return to the browser (Trace ID, error ID, and technical error message).
The technical error message is not really necessary, I think. But the error ID (string enum) is used to show a translated error message to the user. The Trace ID is shown to the user also in several cases.

Since the breaking change in the routes-files-structure, we have three different error mapper functions, that can handle the same Errors in different ways:
- The +server.ts error mapper returns a Response object with the three information as JSON body (which is not a problem)
- The +page.server.ts error mapper for actions returns the `{status: number, errors: Record<string,string>}` object with the three information in the `errors` property
- The +page.server.ts error mapper for the load function calls `throw error()` with a JSON stringified version of the three information

We do not use the `handleError` hook, because we already log every error inside the error mappers (by the way, we use OpenTelemetry JSON logging). We also need to log expected errors, because sometimes a bug causes a false expected error that is actually unexpected.

In the browser, we use the library `svelte-intl-precompile` for i18n. Basically, this library provides a big map in the browser with all translated strings. Each of these strings is identified with a unique id. The error ID we return to the browser is such an ID (yes, it is not really an "error ID" but an "i18n ID", but we just call it so).

---

**Reason for this**

We have to deal with three types of errors, which is why we need error mappers in any server-side endpoint:
1. Explicitly thrown errors (for example validation errors)
2. Thrown errors from outgoing calls (because we use an auto-generated library for outgoing calls, we cannot change this behavior)
3. Anything that is thrown by some other libraries we use (which could be anything from a string to a piece of lasagna).

Since it is not possible to send these totally different types of errors to the browser, we have to map them on the server side into a uniform structure that can be sent to the browser. Currently, we solve this with the custom error mappers mentioned above.

---

**Conclusion**

I think I can break down this whole brain-twisting thing into two basic things:
1. We need a server-side error mapper for +page.server.ts (load and action endpoints) (and for +server.ts, but there we can send whatever Response we want anyways)
2. In the error case, we need a way to pass more than one piece of information to the browser

I think that an `onLoadError` function is currently not necessary for our use case.

My first thought after seeing that `throw error()` only accepts one string as an error message was, why not just pass an object as error message (even it would not be a "message" anymore). But since SvelteKit returns an instance of `Error` to the browser (which I did not know at first) this seems not really possible.

**Reply by Rich-Harris** [reply]
Just spitballing here: what if `handleError` allowed you return `{ metadata }`? That way you could do this sort of thing:

```js
export function handleError({ error }) {
  const { id, traceid } = sendToLoggingService(error);

  return {
    metadata: { id, traceid }
  };
}
```

```svelte
<h1>An error happened</h1>

{#if $page.error.metadata}
  <p class="smallprint">{$page.error.metadata.id}</p>
{/if}
```

Additionally, the 'silent' error helper could be augmented to take an optional third `metadata` argument:

```js
const { id, traceid } = sendToLoggingService(error);
throw error(404, 'not found', { id, traceid });
```

Since validation `errors` have an arbitrary shape, you could continue to do what you like in form actions. (We're in the process of overhauling actions, and it's not 100% clear what we'll land on, but we can be confident that you'd still be able to return whatever metadata you like alongside your actions.)

Obviously we couldn't attach metadata to unexpected client-side errors unless we implemented `onLoadError` or whatever, but it sounds like that wouldn't be useful in your case anyway.

Further thoughts:

* `handleError` could return `{ message, metadata }` instead, giving us the ability to customise the `Internal Error` message. But then you'd be able to do that on the server but not in the browser, which would be weird. Unless we added `onLoadError` and made it behave similarly
* `throw error(status, message, metadata)` could instead be `throw error(status, { message, metadata })` and it would arguably feel more consistent — the POJO you throw is equivalent to `$page.error`
* Which is weirder — having two functions with nearly identical signatures but a different name...
  
    ```js
    // src/hooks.js
    export function handleError({ event, error }) {
      // event is a RequestEvent
    }
    ```

    ```svelte
    <!-- src/routes/+layout.svelte -->
    <script>
      import { onLoadError } from '$app/navigation';

      onLoadError(({ event, error }) => {
        // event is a LoadEvent
      });
    </script>
    ```

    ...or using `handleError` in both cases even though one is an exported function and another is the callback to an imported lifecycle function?

    ```diff
    <!-- src/routes/+layout.svelte -->
    <script>
    -  import { onLoadError } from '$app/navigation';
    +  import { handleError } from '$app/navigation';

    -  onLoadError(({ event, error }) => {
    +  handleError(({ event, error }) => {
        // event is a LoadEvent
      });
    </script>
    ```

**Reply by vekunz** [reply]
> Just spitballing here: what if `handleError` allowed you return `{ metadata }`?

> Additionally, the 'silent' error helper could be augmented to take an optional third `metadata` argument

I think the metadata thing would be a great way to solve our use case.

> Obviously we couldn't attach metadata to unexpected client-side errors unless we implemented `onLoadError` or whatever, but it sounds like that wouldn't be useful in your case anyway.

Currently, we do not use load functions in +page.ts files, so this would not be necessary at least for us (actually I am not sure, where they went because at some point they existed).


> `throw error(status, message, metadata)` could instead be `throw error(status, { message, metadata })` and it would arguably feel more consistent — the POJO you throw is equivalent to `$page.error`

I am not sure which would be better. But the second would be expandable somewhere in the future if needed without a breaking change. But the first one would be easier, I think because then SvelteKit could add some information to `$page.error` itself if needed somewhere in the future. It feels weird if we `throw error(status, { message, metadata })` and SvelteKit adds some more information to it, but I do not know if this is really a case somewhere in the future (or even now).

**Reply by AlexRMU** [reply]
Is it possible to make `handleError` handle an error and return a new one, how does `handle` handle a request and return a response?

**Reply by Rich-Harris** [reply]
> I am not sure which would be better. But the second would be expandable somewhere in the future if needed without a breaking change.

I'm leaning towards `throw error(status, object)` and `handleError` returning the same type of `object`, because there's then a neat symmetry with `$page.error` being the exact same object. (We would need to add the client-side `handleError` equivalent in that case.)

We can also make all this stuff type-able, by having an `App.PageError` type:

```ts
// in your src/app.d.ts
declare namespace App {
  interface PageError {
    message: string;
    metadata: {
      id: string;
      traceid: string;
    }
  }
}
```

One wrinkle with the typing is that the framework default would either need to be this...

```ts
declare namespace App {
  interface PageError {} // allows user to specify any shape, but means a custom `PageError` _must_ be declared
}
```

...or this:

```ts
declare namespace App {
  interface PageError {
    message: string; // sensible default behaviour, but all custom page error types must _also_ include `message: string`
  } 
}
```

> Is it possible to make `handleError` handle an error and return a new one, how does `handle` handle a request and return a response?

Not sure I understand — what would the returned error be for?

**Reply by AlexRMU** [reply]
An uncaught error and/or `error()` will be sent to `handleError`, then there it will be processed/changed and returned, then it will be sent to the client.

**Reply by Rich-Harris** [reply]
But what does 'sent to the client' mean? We have to serialize it, and since we can't do so with perfect fidelity (e.g. if the error uses a custom class, etc), that means we have to figure out a JSON-friendly (and non-sensitive-detail-leaking) representation. Turning an error into that representation is what `handleError` is for, under this proposal.

**Reply by AlexRMU** [reply]
Yes. I mean, `handleError` takes an `uncaught error` and/or `error()` and returns POJO, which will be in `$page.error`. That is, you can return from the `handleError` message, metadata and anything you want.

**Reply by danawoodman** [reply]
I really like the last proposal @Rich-Harris. It allows for flexibility and type safety which is really nice. 

I'm basically doing a workaround now for that in my +server files where I have a custom error "thrower" then I catch them in handleError and strip out the metadata and return a formatted JSON error back to the user. 

It works well but it would be clunky to do in every Kit app. Having most of this built-in would be the bees knees 🐝 🦵

**Reply by vekunz** [reply]
> Yes. I mean, `handleError` takes an `uncaught error` and/or `error()` and returns POJO, which will be in `$page.error`. That is, you can return from the `handleError` message, metadata and anything you want.

I think this is exactly what Rich is talking about.

**Comment by wiverson** [comment]
I would just like to note the existence of RFC 7807 for error handling. I found implementations floating around for a number of other frameworks and it would be really nice to be able to point to an RFC for Svelte/SvelteKit error handling.

https://www.rfc-editor.org/rfc/rfc7807
https://blog.axway.com/learning-center/apis/api-design/introduction-to-rfc-7807

**Reply by wiverson** [reply]
Oh, and here is some more detail about the back & forth around error handling in Spring Boot land: https://github.com/spring-projects/spring-boot/issues/19525

**Reply by Rich-Harris** [reply]
Thanks, I hadn't come across RFC 7807. Do you have any thoughts on how this would translate into an API for SvelteKit?

**Reply by wiverson** [reply]
I think it's basically a matter of having the fields of the error JSON object that gets thrown esp in an API route map to the fields in the RFC. In other words, when SvelteKit generates an error JSON the fields would just match the naming conventions in the RFC.

The RFC is only a few pages and I figured that standardizing on those for how errors are thrown and caught would be a nice, easy interop story.

Basically, it's a few fields at the top level (details in the RFC link):

type, title, status, detail, instance

So, from a high level I think it's just having the default error object match these fields and also avoiding naming conflicts.

If you would like me to write it up more completely I would be happy to do that as an issue or new discussion. Also happy to generate a PR if that would be helpful, tho I'm not an expert on SvelteKit internals/etc. FWIW having gone through a lot of IEEE, W3C RFCs this one is kind of hilariously simple, just tweaking the error json. Off the cuff the only thing that comes to mind is that it might(?) affect tests expecting the current error message format.

**Reply by vekunz** [reply]
I do not think this RFC would be a good decision for SvelteKit. It looks like it is primarily for backend systems (public APIs or distributed systems). But in SvelteKit we have only communication between the SvelteKit server and the browser. In my opinion, this RFC would be way too much for something like SvelteKit.

**Reply by einarpersson** [reply]
> But in SvelteKit we have only communication between the SvelteKit server and the browser.

?? Maybe I’m missing your point, but that totally depends on the application. My SK app has a user interface with data visualizations **and** indeed a public api to expose the underlying data.

(If SK would go in a direction that discourage publishing apis beside UI I’d want to know that because then I’d have to abandon SK and go elsewhere)

(or maybe you are just talking about SK internals, idk. Ignore this then)

**Reply by vekunz** [reply]
Sure, but public API endpoints can only exposed through +server.ts endpoints, and these endpoints do not rely on the SvelteKits error handling.

**Reply by Rich-Harris** [reply]
Yeah, I've been struggling to get my head round what this would look like — it would be unfortunate if (for example) errors had to have a `title` instead of a `message` in order to be compatible with the RFC. But @vekunz nails it — you'd only really need to return these responses from a `+server.js` where you already have complete control, so it's probably easier if we don't have any tighter integration than that.

**Comment by Rich-Harris** [comment]
One possible solution to the client-side error handling question above would be to export a `handleError` hook from the root layout:

```svelte
<!-- src/routes/+layout.svelte -->
<script>
  export function handleError({ event, error }) {
    // `event` is a `LoadEvent`, containing `url`, `params`, `routeId` etc
    // `error` is an unexpected error that was thrown

    return {
      message: `A custom message that isn't just 'Error'`,
      some_other_property: 'whatever'
    };
  }
</script>
```

In effect, `+layout.svelte` becomes a client-side version of `src/hooks.js` (only for `handleError` for now, but if there are other hooks that it makes sense to have client versions of, they could also go here).

The only wrinkle is it wouldn't be able to handle an error that occurred during the initial hydration — that would have to be replaced with a generic error. Otherwise, this seems better than a) introducing a new client-side hooks concept, or b) a lifecycle function with weird semantics

**Reply by Rich-Harris** [reply]
For clarity, it might make sense to rename `src/hooks.js#handleError` to `handleServerError` and the function above to `handleClientError` (and the associated types to `HandleServerError` and `HandleClientError`)

**Reply by dummdidumm** [reply]
I think it's weirder that having a client hooks file, even if it's only that one hook (for now?). Having a client hooks file could also make this work for the initial hydration, since it can be instantiated before the root layout.

**Reply by danawoodman** [reply]
`src/hooks/client.ts` and `src/hooks/server.ts`? I like that

**Reply by Rich-Harris** [reply]
Yeah, we decided to go with `src/hooks/{server,client}.js`

**Reply by syffs** [reply]
This [sample](https://github.com/sveltejs/kit/blob/master/documentation/docs/07-hooks.md) kinda feels uncomplete and misleading:
- are we supposed to use `@sentry/svelte` or something else ?
- it's quite unclear where `Sentry.init` is supposed to be called. I'm calling it in `hooks.server.js`, but not really sure whether there's any impact.

> Yeah, we decided to go with `src/hooks/{server,client}.js`

@Rich-Harris  does this mean `hooks.{server,client}.js` will be soon deprecated ?

**Reply by danawoodman** [reply]
I think its the other way around, they decided to go with hooks.server/hooks.client instead of a directory
